src/IECore/PointsAlgo.cpp

```diff
@@ -297,6 +352,109 @@ PointsPrimitivePtr deletePoints( const PointsPrimitive *pointsPrimitive, const P
 
 }
 
+PointsPrimitivePtr mergePoints( const std::vector<ConstPointsPrimitivePtr> &pointsPrimitives )
+{
+	size_t totalPointCount = 0;
+	typedef std::map<std::string, TypeId > FoundPrimvars;
+	FoundPrimvars foundPrimvars;
+
+	PrimitiveVariableMap constantPrimVars;
+
+	std::vector<PointsPrimitivePtr> validatedPointsPrimitives( pointsPrimitives.size() );
+
+	// find out which primvars can be merged
+	for ( size_t i = 0; i < pointsPrimitives.size(); ++i )
+	{
+		PointsPrimitivePtr pointsPrimitive = validatedPointsPrimitives[i] = pointsPrimitives[i]->copy();
+
+		totalPointCount += pointsPrimitive->getNumPoints();
+		PrimitiveVariableMap& variables = pointsPrimitive->variables;
+		for (PrimitiveVariableMap::iterator it =  variables.begin(); it != variables.end(); ++it)
+		{
+			DataPtr data = it->second.data;
+			TypeId typeId = data->typeId();
+			PrimitiveVariable::Interpolation interpolation = it->second.interpolation;
+			const std::string& name = it->first;
+
+			bool bExistingConstant = constantPrimVars.find( name ) != constantPrimVars.end();
+			FoundPrimvars::const_iterator fIt = foundPrimvars.find( name );
+			bool bExistingVertex = fIt != foundPrimvars.end();
+
+			if (interpolation == PrimitiveVariable::Constant)
+			{
+				if ( bExistingVertex )
+				{
+					std::string msg = boost::str( boost::format( "PointsAlgo::mergePoints mismatching primvar %s") % name );
+					throw InvalidArgumentException( msg );
+				}
+
+				PrimitiveVariableMap::const_iterator constantPrimVarIt = constantPrimVars.find( name );
+				if( !bExistingConstant )
+				{
+					constantPrimVars[name] = it->second;
+				}
+				continue;
+			}
+
+			if (interpolation == PrimitiveVariable::Vertex)
+			{
+
+				PrimitiveVariableMap::const_iterator constantPrimVarIt = constantPrimVars.find( name );
+				if( constantPrimVarIt != constantPrimVars.end() )
+				{
+					std::string msg = boost::str( boost::format( "PointsAlgo::mergePoints mismatching primvar %s" ) % name );
+					throw InvalidArgumentException( msg );
+				}
+
+				if( !bExistingVertex )
+				{
+					foundPrimvars[name] = typeId;
+				}
+				else
+				{
+					if( fIt->second != typeId )
+					{
+						DataCastOp castOp;
```

Comment:
I think we prefer `DataCastOpPtr castOp = new DataCastOp();` though to be honest, I don't remember why...

Reply:
I'd rather not cargo cult this ..

Reply:
It's because DataCastOp is RefCounted, and will therefore delete itself when the count drops to zero - that assumes allocation on the heap. You're getting away with it because nothing is either adding or removing a reference in this case. Really it would be best for this functionality to live in `DataAlgo::cast()` and we wouldn't need an op.

I've been thinking about this a bit lately in relation to a similar problem - making sure that `new RefCountedDerivedClass` is always assigned to a `Ptr` so ownership is taken. What do you think to adding something like `T::Ptr IECore::makeCounted<T>( Args&&... args )`, which would be analogous to `std::make_shared` and `std::make_unique`? If we adopted that across the board then in a later version we could make the RefCounted constructors private, with makeCounted being a friend, and we'd be enforcing correct usage.

Reply:
Thanks for the notes - I'll change to a heap object. I assumed the ref-counting was fine as there wasn't an exception on dtor and didn't consider this incorrect usage.

I like the idea of a single makeCounted function to ensure stack ref counted objects can't be created to enforce the correct usage.

Reply:
> I assumed the ref-counting was fine as there wasn't an exception on dtor and didn't consider this incorrect usage.

Yeah, it's a bit debatable for sure. I'm pretty sure I've made an Op on the stack before too, because it seemed so daft to have to allocate one, but the reality is that it's still going to allocate parameters and things behind the scenes anyway, so you're not really saving much. I think the real issue is that we need the same dataCast functionality available as a simple function and all this becomes moot.

Reply:
Are you suggesting that I make the DataAlgo::cast function now as part of this PR?

Reply:
No, not at all. Just mentioning stuff I think we should be thinking of for the future...